nyx: attempt at creating a propagation method

gwbres opened the pull request.

ChristopherRabotin commented:
I'll provide a more detailed answer tomorrow, so I've just made a few comments on the code for now.

ChristopherRabotin commented:
Great! That's probably what we should use because Nyx requires a fully defined orbit (position, velocity, epoch, frame) to start a propagation.

gwbres commented:
@ChristopherRabotin,

that's truly awesome 👍 but man I did not expect that struggle 😄 
It's almost impossible to find SP3 with dynamics. 
And it turned out I only had the 5-line example that IGS provides with the SP3 specs.
So.. quite far from meaningful for this topic.

Went out browsing for a while. Found `https://earth-info.nga.mil` that offers a basic downloading interface (GNSS tab), with 9 or 30 day prediction download. It turns out those are revA files, and we do not support that well. I manually patched the files so we can parse them and get started. I just pushed one of the 9 files I downloaded. Note that those are predicted orbits (assumed from the website name), so they are not released by the IGS POD. So it's basically their propagation model (whatever it might be, and we have no clue) against yours.

The test streams out zeros only at the moment. I saw the prediction flag arise where it should (but I already validated that bit in the last release). I'll investigate tomorrow, now I just need pasta and some sleep 

Patch
====

I reworked the patch so it may apply to _all_ satellites, the idea is to be able to propagate an entiere fileset, so we can dump it.
After the prediction, the SP3 structure is updated so the it reflects when we dump it.

I renamed all of these features with `fn spatial_()`, and that's important. Follwoing #24 and #26, obviously spatial and temporal are two different dimensions, but SP3 allows describing a prediction for both of them. So I hope one day to propose a temporal prediction method, and I think researching this topic will be _very_ interesting.

> Great! That's probably what we should use because Nyx requires a fully defined orbit (position, velocity, epoch, frame) to start a propagation.

Can't we foresee an OD without dynamics at all ? it turns out to truly be an issue.

Thoughts / Q&A
===========

0. could we come up with some kind of preprocessing to apply to Self and determine the dynamics, simply derivating all the previously parsed states, so we could deploy the OD on _any_ SP3 ? considering how difficult it is to find SP3 with dynamics.
It also turns out that, if we can offer dynamics determination it could be quite useful.

1. If your orbital fitter can provide some kind of fit deviation information, it is possible to format it in SP3. (although, not supported by the currently version, these columns are disregarded). For example, all `XXXIGSDEMOSXXXX` files on their FTP has deviations

![image](https://github.com/user-attachments/assets/826d60f9-a192-437e-bd06-cbfdfd1e7e26)

2. if we're ever able to make sense of the [OrbitType](https://github.com/rtk-rs/sp3/blob/main/src/header/mod.rs#L57) described in the header, it may relate to your fitting method(s) and so, could also provide meaningful information. The source of information for all of this, is the IGS FTP server.

3. If we ever understand the algorithms th IGS labs use to do their own fit, you can compare their predictions to your predictions. For example, `data/SP3/C/ESA0OPSULT_20232320` is terminated by predicted states, so direct output from their algorithm. So quite another interesting aspect.

gwbres commented on src/nyx.rs:
```diff
@@ -0,0 +1,251 @@
+        // create a propagator for each satellite
+        // using the same solar pressure and dynamics model
+        let iter = self
+            .spacecraft_model_iter(initial_epoch)
```
@ChristopherRabotin,

so we initiate the spacecraft model using the specified point in time.
It's not exactly how I thought it would work, are we sure we don't have better things to do here ? let's say this is noon and we have 12 hours of IGS garanteed states. Can't we integrate all those past states somehow, and so, reduce the propagation error ? instead of only taking the latest state into account

ChristopherRabotin commented on src/nyx.rs:
```diff
@@ -0,0 +1,251 @@
+    ) -> Result<Box<dyn Iterator<Item = (SV, Traj<Spacecraft>)> + '_>, PredictionError> {
+        let orbital_model = OrbitalDynamics::point_masses(vec![MOON, SUN]);
+
+        let srp_model = SolarPressure::new(vec![EARTH_J2000, MOON_J2000], almanac.clone())?;
```
```suggestion
        let srp_model = SolarPressure::new(vec![EARTH_J2000], almanac.clone())?;
```

We can most likely disable the shadow casted from the Moon here. It may happen once every few months, but not enough to significantly alter the propagation of the spacecraft.

ChristopherRabotin commented:
One easy way to get the velocity data is to subtract the position of two SP3 positions: what is the typical time step of an SP3 file? If there is one data point every few seconds, I think that will give good enough initial guess for the velocity. We can assess after running the test whether that is a sufficient approach.

gwbres commented:
> One easy way to get the velocity data is to subtract the position of two SP3 positions: what is the typical time step of an SP3 file? If there is one data point every few seconds, I think that will give good enough initial guess for the velocity. We can assess after running the test whether that is a sufficient approach.

the sampling period of the files released is normally 15' or 5', 15' being the most common.
And then the users interpolate the samples on their side.
So maybe we should interpolate around the initial_state, calculate the velocity at that time and then deploy

gwbres commented on src/nyx.rs:
```diff
@@ -0,0 +1,108 @@
+
+        for (k, v) in final_states.iter() {
+            // Deploy a process
+            let mut odp = SpacecraftKalmanOD::new(
```
does that work backwards also ? so actually predict in the past.
Maybe `start` + `duration` is not enough, I should simply have the user define its own `TimeSerie`, that's what it was made for

ChristopherRabotin replied:
Yes, you can propagate with a negative duration or an epoch that is before your current state.

gwbres commented:
Let me close a first issues today (#30) that will help this topic.
Once we have it, I can then run the comparison on a standard SP3 (IGS and not fully predicted).
 But standard having 15' sample rate, I'd like to hear your thoughts on simply deriving the dynamics from that

ChristopherRabotin commented:
Hmm, OK. Can you give me an initial state with epoch, and I can debug it later today?

gwbres commented:
> Hmm, OK. Can you give me an initial state with epoch, and I can debug it later today?

I added a debug trace for each spacecraft object that I initialized. This is my "forward_prop" test, so we can see that the structure is initialized at the correct timestamp (noon for that day) and so (read following) the dynamics are resolved. Everything looks good to me. Because my spacecraft initialization is very minimalist, everything is pretty much default, including the SC state "coast", is that important ?

[nyx-debug.log](https://github.com/user-attachments/files/21202916/nyx-debug.log)

Thanks, that made me realize I have to be careful with the initialization instant, when the dynamics have been manually derived. Because in this workflow, it is impossible to derive the dynamics for the very first epoch. For example, in my "backwards" test, starting at midnight and propagating backwards for a day, that should not work correctly either. In the end, I think the dynamics derivation method will have to be more complex than this basic draft

ChristopherRabotin commented:
The spacecraft should be in Coast: the other mode is only for maneuver design for low thrust propulsion.

Surprisingly, I can't seem to recreate the issue... Here is the test I added to Nyx to rebuild the first error from the log -- can you let me know if you notice any error.

```diff

+#[rstest]
+fn test_rtk(almanac: Arc<Almanac>) {
+    use std::str::FromStr;
+    let _ = pretty_env_logger::try_init();
+
+    let epoch = Epoch::from_str("2020-06-25T12:00:00 GPST").unwrap();
+    let eme2k = almanac.frame_from_uid(EARTH_J2000).unwrap();
+    let start_state = Orbit::cartesian(
+        10996.104343,
+        -19841.200560,
+        -13758.983598,
+        1.620568,
+        -0.922680,
+        2.398326,
+        epoch,
+        eme2k,
+    );
+
+    println!("{start_state}\n{start_state:x}");
+    let setup = Propagator::default_dp78(SpacecraftDynamics::new(OrbitalDynamics::point_masses(
+        vec![MOON, SUN],
+    )));
+    let final_state = setup
+        .with(start_state.into(), almanac.clone())
+        .for_duration(Unit::Hour * 12)
+        .unwrap();
+    println!("{final_state}\n{final_state:x}");
+}
+
```

And the result:

```
running 1 test
[Earth J2000] 2020-06-25T12:00:00 GPST	position = [10996.104343, -19841.200560, -13758.983598] km	velocity = [1.620568, -0.922680, 2.398326] km/s
[Earth J2000] 2020-06-25T12:00:00 GPST	sma = 19146.649970 km	ecc = 0.387335	inc = 74.142040 deg	raan = 308.916767 deg	aop = 150.903171 deg	ta = 176.473297 deg
 INFO  nyx_space::propagators::instance > Propagating for 12 h until 2020-06-26T00:00:00 GPST
 INFO  nyx_space::propagators::instance > 	... done in 304 ms 746 μs 231 ns
total mass = 0.000 kg @  [Earth J2000] 2020-06-26T00:00:00 GPST	position = [-8351.198443, 4740.484039, -12390.315892] km	velocity = [1.013159, -3.210532, -4.324660] km/s  Coast
total mass = 0.000 kg @  [Earth J2000] 2020-06-26T00:00:00 GPST	sma = 19146.530005 km	ecc = 0.387332	inc = 74.141426 deg	raan = 308.916823 deg	aop = 150.906495 deg	ta = 84.346617 deg  Coast
test propagation::trajectory::test_rtk ... ok
```

The only difference I can tell is that I'm not currently turning on the SRP: it should be turned on. I don't see where you set the SRP area or Cr value for the spacecraft. A good Cr value should be the default of `1.2`. The area should be the area that will be pointing to the Sun. In real world applications, I've never had to use more than an approximate number for both: the OD process _may_ be able to converge on a good Cr (depending on the orbital regime), but the area is the biggest contributor.

Could you tell me what are the area and Cr values you used for these space vehicles? I'll update my test and check again.

Thanks
